### PR TITLE
fix(connlib): prioritize dns resources to CIDR ones in case of an overlap

### PR DESCRIPTION
**Pull request by conectado**
For full route this happens always and if we don't prioritize DNS resources any packet for DNS IPs will get routed to the full route gateway which might not have the correct resource.

TODO: this still needs unit tests
TODO: Waiting on #5891

### CODE REVIEW

**Review by thomaseizinger — COMMENTED** [review]
I guess this would otherwise only happen if somebody defines a CIDR resource that is within the subnet for which we generate dummy DNS IPs?

Should there be a warning in the portal for such resources? They are effectively unroutable then, right? Unless the CIDR resource happens to be globally addressable (which is unlikely) or the DNS resource happens to be the same?

**Comment by conectado** [comment]
> I guess this would otherwise only happen if somebody defines a CIDR resource that is within the subnet for which we generate dummy DNS IPs?
> 
> Should there be a warning in the portal for such resources? They are effectively unroutable then, right? Unless the CIDR resource happens to be globally addressable (which is unlikely) or the DNS resource happens to be the same?
 
With @jamilbk  we were planning to restrict that in the portal.

**Comment by conectado** [comment]
> > I guess this would otherwise only happen if somebody defines a CIDR resource that is within the subnet for which we generate dummy DNS IPs?
> > Should there be a warning in the portal for such resources? They are effectively unroutable then, right? Unless the CIDR resource happens to be globally addressable (which is unlikely) or the DNS resource happens to be the same?
> 
> With @jamilbk we were planning to restrict that in the portal.

To elaborate further:

If someone added such a CIDR resource it'd work quite weirdly! With this change it would be routed to the cidr resource until the route is used as a DNS at which point it will change.

Though, something weird is that this case will happen with full-route-- maybe we need to drop those packets until the DNS route is added.

**Comment by jamilbk** [comment]
See #5844

**Review by jamilbk — APPROVED** [review]

**Comment by thomaseizinger** [comment]
@conectado The changes you are making here make the proptests even faster! Finishes now in 15s instead fo 20 in debug mode on my laptop! With optimisations it is only 1.5s :heart_eyes:

**Comment by thomaseizinger** [comment]
> @conectado The changes you are making here make the proptests even faster! Finishes now in 15s instead fo 20 in debug mode on my laptop! With optimisations it is only 1.5s 😍

This is becoming truly incredible!

**Comment by conectado** [comment]
Still trying to figure this tests out! I need to know in the reference state what the DNS ips are since a "cidr packet" might trigger a connection for a dns resource now(or a packet flow)

**Comment by conectado** [comment]
> For full route this happens always and if we don't prioritize DNS resources any packet for DNS IPs will get routed to the full route gateway which might not have the correct resource.
> 
> TODO: this still needs unit tests
TODO: Merge after #5840

**Comment by thomaseizinger** [comment]
> Still trying to figure this tests out! I need to know in the reference state what the DNS ips are since a "cidr packet" might trigger a connection for a dns resource now(or a packet flow)

Hmm, maybe we should skip testing this. In fact, the way we should probably test this is to generate arbitrary CIDR ranges but explicitly filter out our reserved ranges to make sure we never generate CIDR traffic for that range. That way, the test documents that this isn't supported right now.

Sorry for sending you on this wild-goose chase!

**Comment by conectado** [comment]
> > Still trying to figure this tests out! I need to know in the reference state what the DNS ips are since a "cidr packet" might trigger a connection for a dns resource now(or a packet flow)
> 
> Hmm, maybe we should skip testing this. In fact, the way we should probably test this is to generate arbitrary CIDR ranges but explicitly filter out our reserved ranges to make sure we never generate CIDR traffic for that range. That way, the test documents that this isn't supported right now.
> 
> Sorry for sending you on this wild-goose chase!

No worries good things came from this! Will do that for this PR so we can merge this, but we might want to re-think this to be able to test full-route.

**Comment by thomaseizinger** [comment]
> > > Still trying to figure this tests out! I need to know in the reference state what the DNS ips are since a "cidr packet" might trigger a connection for a dns resource now(or a packet flow)
> > 
> > 
> > Hmm, maybe we should skip testing this. In fact, the way we should probably test this is to generate arbitrary CIDR ranges but explicitly filter out our reserved ranges to make sure we never generate CIDR traffic for that range. That way, the test documents that this isn't supported right now.
> > Sorry for sending you on this wild-goose chase!
> 
> No worries good things came from this! Will do that for this PR so we can merge this, but we might want to re-think this to be able to test full-route.

For full route, the packets do get routed to the gateway so we are good. All we need to change is a flag in the reference state to know whether full route resource is active :)

**Comment by conectado** [comment]
> > > > Still trying to figure this tests out! I need to know in the reference state what the DNS ips are since a "cidr packet" might trigger a connection for a dns resource now(or a packet flow)
> > > 
> > > 
> > > Hmm, maybe we should skip testing this. In fact, the way we should probably test this is to generate arbitrary CIDR ranges but explicitly filter out our reserved ranges to make sure we never generate CIDR traffic for that range. That way, the test documents that this isn't supported right now.
> > > Sorry for sending you on this wild-goose chase!
> > 
> > 
> > No worries good things came from this! Will do that for this PR so we can merge this, but we might want to re-think this to be able to test full-route.
> 
> For full route, the packets do get routed to the gateway so we are good. All we need to change is a flag in the reference state to know whether full route resource is active :)

I see, we handle it as a special transition of adding the internet resource instead of adding the full-route cidr resource :+1:

**Comment by conectado** [comment]
> > > > > Still trying to figure this tests out! I need to know in the reference state what the DNS ips are since a "cidr packet" might trigger a connection for a dns resource now(or a packet flow)
> > > > 
> > > > 
> > > > Hmm, maybe we should skip testing this. In fact, the way we should probably test this is to generate arbitrary CIDR ranges but explicitly filter out our reserved ranges to make sure we never generate CIDR traffic for that range. That way, the test documents that this isn't supported right now.
> > > > Sorry for sending you on this wild-goose chase!
> > > 
> > > 
> > > No worries good things came from this! Will do that for this PR so we can merge this, but we might want to re-think this to be able to test full-route.
> > 
> > 
> > For full route, the packets do get routed to the gateway so we are good. All we need to change is a flag in the reference state to know whether full route resource is active :)
> 
> I see, we handle it as a special transition of adding the internet resource instead of adding the full-route cidr resource 👍

Well, we will still have to exclude packets to "full route" that are in the DNS range, so that will go untested.

File: rust/connlib/tunnel/src/tests/reference.rs [hunk 1]
```diff
@@ -132,9 +136,22 @@ impl ReferenceStateMachine for ReferenceState {
             )
             .with(
                 1,
-                (cidr_resource(8), sample::select(state.all_gateways())).prop_map(
-                    |(resource, gateway)| Transition::AddCidrResource { resource, gateway },
-                ),
+                (cidr_resource(8), sample::select(state.all_gateways()))
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/tests/reference.rs]: I think at this complexity level, we may want to extract a dedicated strategy function now. Also allows us to make a comment explaining this.

File: rust/connlib/tunnel/src/tests/reference.rs [hunk 2]
```diff
@@ -132,9 +136,22 @@ impl ReferenceStateMachine for ReferenceState {
...
+                    .prop_filter(
+                        "tests doesn't support yet CIDR resources overlapping DNS resources",
+                        |(r, _)| {
+                            IpNetwork::from_str(IPV4_RESOURCES)
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/tests/reference.rs]: Didn't we at some point upstream a const-ctor?

**Review by thomaseizinger — APPROVED** [review]

File: rust/connlib/tunnel/src/tests/reference.rs (hunk 2 above)
Comment (conectado) [reply rust/connlib/tunnel/src/tests/reference.rs]: I'm waiting on [this](https://github.com/JakubOnderka/ip_network/issues/13), we could patch with `main`(`ip_network_table` uses crates.io version)

**Pull request merged**